yt_dlp/extractor/goplay_auth_aws.py

```diff
@@ -0,0 +1,317 @@
+                     '670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B' + \
+                     'E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9' + \
+                     'DE2BCBF6955817183995497CEA956AE515D2261898FA0510' + \
+                     '15728E5A8AAAC42DAD33170D04507A33A85521ABDF1CBA64' + \
+                     'ECFB850458DBEF0A8AEA71575D060C7DB3970F85A6E1E4C7' + \
+                     'ABF5AE8CDB0933D71E8C94E04A25619DCEE3D2261AD2EE6B' + \
+                     'F12FFA06D98A0864D87602733EC86A64521F2B18177B200C' + \
+                     'BBE117577A615D6C770988C0BAD946E208E24FA074E5AB31' + \
+                     '43DB5BFCE0FD108E4B82D120A93AD2CAFFFFFFFFFFFFFFFF'
+
+        # https://github.com/aws/amazon-cognito-identity-js/blob/master/src/AuthenticationHelper.js#L49
+        self.g_hex = '2'
+        self.info_bits = bytearray('Caldera Derived Key', 'utf-8')
+
+        self.big_n = self.__hex_to_long(self.n_hex)
+        self.g = self.__hex_to_long(self.g_hex)  # pylint: disable=invalid-name
+        self.k = self.__hex_to_long(self.__hex_hash('00' + self.n_hex + '0' + self.g_hex))  # pylint: disable=invalid-name
+        self.small_a_value = self.__generate_random_small_a()
+        self.large_a_value = self.__calculate_a()
+
+    def authenticate(self, username, password):
+        """ Authenticate with a username and password. """
+        # Step 1: First initiate an authentication request
+        auth_data_dict = self.__get_authentication_request(username)
+        auth_data = json.dumps(auth_data_dict).encode("utf-8")
+        auth_headers = {
+            "X-Amz-Target": "AWSCognitoIdentityProviderService.InitiateAuth",
+            "Accept-Encoding": "identity",
+            "Content-Type": "application/x-amz-json-1.1"
+        }
+        auth_request = compat_urllib_request.Request(url=self.url, data=auth_data, headers=auth_headers)
+        auth_response = compat_urllib_request.urlopen(auth_request).read()
+        auth_response_json = json.loads(auth_response.decode("utf-8"))
+        challenge_parameters = auth_response_json.get("ChallengeParameters")
+
+        challenge_name = auth_response_json.get("ChallengeName")
+        if not challenge_name == "PASSWORD_VERIFIER":
+            raise AuthenticationException(auth_response_json.get("message"))
+
+        # Step 2: Respond to the Challenge with a valid ChallengeResponse
+        challenge_request = self.__get_challenge_response_request(challenge_parameters, password)
+        challenge_data = json.dumps(challenge_request).encode("utf-8")
+        challenge_headers = {
+            "X-Amz-Target": "AWSCognitoIdentityProviderService.RespondToAuthChallenge",
+            "Content-Type": "application/x-amz-json-1.1"
+        }
+        auth_request = compat_urllib_request.Request(url=self.url, data=challenge_data, headers=challenge_headers)
+        auth_response = compat_urllib_request.urlopen(auth_request).read()
```

Comment:
this should use `ie._download_webpage` etc. You could pass an instance of the extractor to the class to access these

Reply:
I've moved the code from `goplay_auth_aws.py` into this file now. I would actually prefer not to update the code itself too much as it works now and resembles the original source(s) as much as possible. Would it be okay to merge it as is?

Reply:
Okay, I see that the tests are now failing after merging the code because of using deprecated helpers from `compat`.

I'll look into it (at another time). 👍

Reply:
>  I would actually prefer not to update the code itself too much as it works now and resembles the original source(s) as much as possible. Would it be okay to merge it as is?

No. The wrappers exist for a reason. Not using them would break various functionality (eg: --proxy)

Reply:
Okay, I finally had some time to work on this. I've replaced all occurrences of `compat_urllib_request.Request` by `_download_webpage` and have gotten rid of the `compat_urllib_request` import.

Does it suffice like this?